File: linkis-commons/linkis-storage/src/main/scala/org/apache/linkis/storage/source/AbstractFileSource.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.storage.source

import org.apache.linkis.common.io.{FsWriter, MetaData, Record}

import org.apache.commons.io.IOUtils
import org.apache.commons.math3.util.Pair

import java.util

import scala.collection.JavaConverters._

abstract class AbstractFileSource(var fileSplits: Array[FileSplit]) extends FileSource {

  override def shuffle(s: Record => Record): FileSource = {
    fileSplits.foreach(_.shuffler = s)
    this
  }

  override def page(page: Int, pageSize: Int): FileSource = {
    fileSplits.foreach(_.page(page, pageSize))
    this
  }

  override def addParams(params: util.Map[String, String]): FileSource = {
    fileSplits.foreach(_.addParams(params))
    this
  }

  override def addParams(key: String, value: String): FileSource = {
    fileSplits.foreach(_.addParams(key, value))
    this
  }

  override def getFileSplits: Array[FileSplit] = this.fileSplits

  override def getParams: util.Map[String, String] =
    fileSplits.map(_.params.asScala).foldLeft(Map[String, String]())(_ ++ _).asJava

  override def write[K <: MetaData, V <: Record](fsWriter: FsWriter[K, V]): Unit =
    fileSplits.foreach(_.write(fsWriter))

  override def close(): Unit = this.fileSplits.foreach(IOUtils.closeQuietly)

  override def collect(): Array[Pair[Object, util.ArrayList[Array[String]]]] =
    fileSplits.map(_.collect())

  override def getTotalLine: Int = this.fileSplits.map(_.totalLine).sum

  override def getTypes: Array[String] = this.fileSplits.map(_.`type`)

  override def getFileInfo(needToCountRowNumber: Int = 5000): Array[Pair[Int, Int]] =
    fileSplits.map(_.getFileInfo(needToCountRowNumber))

}

File: linkis-commons/linkis-storage/src/main/java/org/apache/linkis/storage/source/AbstractFileSource.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.storage.source;

import org.apache.linkis.common.io.FsWriter;
import org.apache.linkis.common.io.MetaData;
import org.apache.linkis.common.io.Record;

import org.apache.commons.io.IOUtils;
import org.apache.commons.math3.util.Pair;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public abstract class AbstractFileSource implements FileSource {

  private FileSplit[] fileSplits;

  public AbstractFileSource(FileSplit[] fileSplits) {
    this.fileSplits = fileSplits;
  }

  @Override
  public FileSource shuffle(Function<Record, Record> function) {
    Arrays.stream(fileSplits).forEach(fileSplit -> fileSplit.shuffler = function);
    return this;
  }

  @Override
  public FileSource page(int page, int pageSize) {
    Arrays.stream(fileSplits).forEach(fileSplit -> fileSplit.page(page, pageSize));
    return this;
  }

  @Override
  public FileSource addParams(Map<String, String> params) {
    Arrays.stream(fileSplits).forEach(fileSplit -> fileSplit.addParams(params));
    return this;
  }

  @Override
  public FileSource addParams(String key, String value) {
    Arrays.stream(fileSplits).forEach(fileSplit -> fileSplit.addParams(key, value));
    return this;
  }

  @Override
  public FileSplit[] getFileSplits() {
    return this.fileSplits;
  }

  @Override
  public Map<String, String> getParams() {
    return Arrays.stream(fileSplits)
        .map(FileSplit::getParams)
        .flatMap(map -> map.entrySet().stream())
        .collect(
            Collectors.toMap(
                Map.Entry::getKey, Map.Entry::getValue, (existingValue, newValue) -> newValue));
  }

  @Override
  public <K extends MetaData, V extends Record> void write(FsWriter<K, V> fsWriter) {
    Arrays.stream(fileSplits).forEach(fileSplit -> fileSplit.write(fsWriter));
  }

  @Override
  public void close() {
    Arrays.stream(fileSplits).forEach(IOUtils::closeQuietly);
  }

  @Override
  public Pair<Object, List<String[]>>[] collect() {
    return Arrays.stream(fileSplits).map(FileSplit::collect).toArray(Pair[]::new);
  }

  @Override
  public int getTotalLine() {
    return Arrays.stream(fileSplits).mapToInt(FileSplit::getTotalLine).sum();
  }

  @Override
  public String[] getTypes() {
    return Arrays.stream(fileSplits).map(FileSplit::getType).toArray(String[]::new);
  }

  @Override
  public Pair<Integer, Integer>[] getFileInfo(int needToCountRowNumber) {
    return Arrays.stream(fileSplits)
        .map(fileSplit -> fileSplit.getFileInfo(needToCountRowNumber))
        .toArray(Pair[]::new);
  }
}
